data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/RenameKeyProcessor.java

```diff
@@ -59,14 +67,28 @@ public Collection<Record<Event>> doExecute(final Collection<Record<Event>> recor
                         continue;
                     }
 
-                    if (entry.getFromKey().equals(entry.getToKey()) || !recordEvent.containsKey(entry.getFromKey())) {
+                    if (Objects.nonNull(entry.getFromKey()) && (entry.getFromKey().equals(entry.getToKey()) || !recordEvent.containsKey(entry.getFromKey()))) {
                         continue;
                     }
-
                     if (!recordEvent.containsKey(entry.getToKey()) || entry.getOverwriteIfToKeyExists()) {
-                        final Object source = recordEvent.get(entry.getFromKey(), Object.class);
-                        recordEvent.put(entry.getToKey(), source);
-                        recordEvent.delete(entry.getFromKey());
+                        if(Objects.nonNull(entry.getFromKey())) {
+                            final Object source = recordEvent.get(entry.getFromKey(), Object.class);
+                            recordEvent.put(entry.getToKey(), source);
+                            recordEvent.delete(entry.getFromKey());
+                        }
+                        if(Objects.nonNull(entry.getFromKeyCompiledPattern())) {
+                            Map<String,Object> eventMap = recordEvent.toMap();
+                            Pattern fromKeyCompiledPattern = entry.getFromKeyCompiledPattern();
+                            for (Map.Entry<String, Object> eventEntry : eventMap.entrySet()) {
+                                final String key = eventEntry.getKey();
+                                final Object value = eventEntry.getValue();
+                                if (fromKeyCompiledPattern.matcher(key).matches()) {
+                                    recordEvent.put(entry.getToKey(), value);
+                                    recordEvent.delete(key);
```

Comment:
When I test the processor with the example data in #4849. I got error from this line because the `|` in the key:
```
IllegalArgumentException: key delivered_at|d87e56e8-f52f-474f-ad18-155b2a08f680 must contain only alphanumeric chars with .-_@/ and must follow JsonPointer (ie. 'field/to/key')
```
It's a validation on keys from JacksonEventKey https://github.com/opensearch-project/data-prepper/blob/00b10bdcb9783bf8e8de20bf95343b9ff4ed77b3/data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/JacksonEventKey.java#L142. We may need to change the validation, at least for the delete method.